pandas/tests/frame/test_timeseries.py

```diff
@@ -57,6 +57,29 @@ def test_diff(self):
             1), 'z': pd.Series(1)}).astype('float64')
         assert_frame_equal(result, expected)
 
+    @pytest.mark.parametrize('axis', [0, 1])
+    @pytest.mark.parametrize('tz', [None, 'UTC'])
```

Comment:
can we separate this into 2 tests, one for axis 0 and one for axis 1. The reason is that on axis 1 need to also test for *different* timezone diffs (which should fail). I think you are going to raise NotImplementedError, but this should really be a TypeError.

Reply:
`axis=1` currently doesn't work even for the same timezone:

```
In [2]: df = DataFrame({0: date_range('2010', freq='D', periods=2, tz='utc'),
   ...:                 1: date_range('2010', freq='D', periods=2, tz='utc')})
   ...:

In [3]: df.diff(axis=1) # incorrect, commented out the NotImplementedError
Out[3]:
       0      1
0    NaT    NaT
1 1 days 1 days

In [4]: df.diff(axis=0) # correct
Out[4]:
       0      1
0    NaT    NaT
1 1 days 1 days
```
since diff operates per block via apply. As you mentioned before this probably needs a whole new implementation in general, but that's why I thought of raising a `NotImplementedError` for `diff(axis=1)` (for now)